sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1.java

```diff
@@ -0,0 +1,620 @@
+      return new BatchWrite.Builder();
+    }
+  }
+
+  /**
+   * Concrete class representing a {@link PTransform}{@code <}{@link PCollection}{@code <}{@link
+   * com.google.firestore.v1.Write}{@code >, }{@link PDone}{@code >} which will write to Firestore.
+   * <p/>
+   * If an error is encountered while trying to write to Cloud Firestore a {@link
+   * FailedWritesException} will be thrown. If you would like a failed write to not result in a
+   * {@link FailedWritesException}, you can instead use {@link BatchWriteWithDeadLetterQueue} which
+   * will instead output any failed write. {@link BatchWriteWithDeadLetterQueue } can be used by
+   * including {@link Builder#withDeadLetterQueue()} when constructing the write handler.
+   * <p/>
+   * This class is part of the Firestore Connector DSL, it has a type safe builder accessible via
+   * {@link FirestoreIO#v1()}{@code .}{@link FirestoreV1#write() write()}{@code .}{@link
+   * FirestoreV1.Write#batchWrite() batchWrite()}.
+   * <p/>
+   * All request quality-of-service for an instance of this PTransform is scoped to the worker and
+   * configured via {@link Builder#withRpcQosOptions(RpcQosOptions)}.
+   * <p/>
+   * Writes performed against Firestore will be ordered and grouped to maximize throughput while
+   * maintaining a high request success rate. Batch sizes will be determined by the QOS layer.
+   *
+   * @see FirestoreIO#v1()
+   * @see FirestoreV1#write()
+   * @see FirestoreV1.Write#batchWrite()
+   * @see BatchWrite.Builder
+   * @see BatchWriteWithDeadLetterQueue
+   * @see BatchWriteRequest
+   * @see com.google.firestore.v1.BatchWriteResponse
+   * @see <a target="_blank" rel="noopener noreferrer" href="https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.Firestore.BatchWrite">google.firestore.v1.Firestore.BatchWrite</a>
+   * @see <a target="_blank" rel="noopener noreferrer" href="https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.BatchWriteRequest">google.firestore.v1.BatchWriteRequest</a>
+   * @see <a target="_blank" rel="noopener noreferrer" href="https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.BatchWriteResponse">google.firestore.v1.BatchWriteResponse</a>
+   */
+  public static final class BatchWrite
+      extends Transform<
+      PCollection<com.google.firestore.v1.Write>,
+      PDone,
+      BatchWrite,
+      BatchWrite.Builder
+      > {
+
+    private BatchWrite(JodaClock clock,
+        FirestoreStatefulComponentFactory firestoreStatefulComponentFactory, RpcQosOptions rpcQosOptions) {
+      super(clock, firestoreStatefulComponentFactory, rpcQosOptions);
+    }
+
+    @Override
+    public PDone expand(PCollection<com.google.firestore.v1.Write> input) {
+      input.apply("batchWrite", ParDo.of(new DefaultBatchWriteFn(clock, firestoreStatefulComponentFactory, rpcQosOptions, CounterFactory.DEFAULT)));
```

Comment:
Are writes to Firestore idempotent ? Note that bundles of this write step may fail and may be retried by the runner. So if the writes are not idempotent you'll have to handle this to prevent duplicate data from being written.

Reply:
Writes to firestore are generally idempotent, there are a few narrow cases where a precondition can be specified on an individual write (something like perform and update as long as the updatedAt matches timeX). If a precondition fails, firestire will respond with a failure status for that individual write, and if the dead letter queue is used will be output as an individual `WriteFailure{write, writeResult, status}`